### PR TITLE
PIN-4254 Save agreements events without organizationId

### PR DESCRIPTION
**Pull request by nttdata-rtorsoli**

### CODE REVIEW

File: src/main/scala/it/pagopa/interop/notifier/service/converters/AgreementEventsConverter.scala [hunk 1]
```diff
@@ -13,26 +13,30 @@ object AgreementEventsConverter {
...
-    case AgreementSuspended(a)                   => Future.successful(MessageId(a.id, a.producerId.toString()).some)
-    case AgreementDeactivated(a)                 => Future.successful(MessageId(a.id, a.producerId.toString()).some)
-    case AgreementContractAdded(id, _)           => getMessageIdFromDynamo(dynamoService)(id).map(_.some)
+  ): Future[Seq[MessageId]] = event match {
+    case AgreementAdded(a)                       => 
+      val messageId = MessageId(a.id, a.producerId.toString())
+      val messageAgreements = MessageId(a.id, agreements)
+      for {
+      _ <- dynamoService.put(messageId)
+      _ <-  dynamoService.put(messageAgreements)
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/notifier/service/converters/AgreementEventsConverter.scala]: unfortunately I think this would not work.
`a.id` is used as key in dynamo table, so the second `put` overrides the first.

I have a solution in mind but let me sleep on it and see if it still makes sense tomorrow

Comment (nttdata-rtorsoli) [reply src/main/scala/it/pagopa/interop/notifier/service/converters/AgreementEventsConverter.scala]: We can use the second parameter as a sort key or we can use id#<agreements or producerId> as partition key

File: src/main/scala/it/pagopa/interop/notifier/api/impl/EventsServiceApiImpl.scala [hunk 2]
```diff
@@ -39,7 +40,10 @@ final class EventsServiceApiImpl(dynamoNotificationService: DynamoNotificationSe
     val operationLabel = s"Retrieving $limit messages from id $lastEventId for partition "
     logger.info(operationLabel)
 
-    val result: Future[Events] = getEvents(agreements, limit, lastEventId)
+    val result: Future[Events] = for {
+      organizationId <- getOrganizationIdFutureUUID(contexts)
+      events         <- getEvents(Seq(organizationId.toString, agreements).mkString(separator), limit, lastEventId)
+    } yield events
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/notifier/api/impl/EventsServiceApiImpl.scala]: This is enough 👍 
```suggestion
    val result: Future[Events] = getEvents(agreements, limit, lastEventId)
```

**Review by galales — COMMENTED** [review]
a couple of changes required:
1. [here](https://github.com/pagopa/interop-be-notifier/blob/1.0.x/src/main/scala/it/pagopa/interop/notifier/model/NotificationMessage.scala#L63) we need to split the `organizationId` and create one message for each resulting value
2. [here](https://github.com/pagopa/interop-be-notifier/blob/1.0.x/src/main/scala/it/pagopa/interop/notifier/service/impl/QueueHandler.scala#L65) we need to split the `organizationId` and basically execute the for-comprehension for each one

File: src/main/scala/it/pagopa/interop/notifier/model/NotificationMessage.scala [hunk 3]
```diff
@@ -51,25 +52,26 @@ object NotificationMessage {
 
   implicit val formatNotificationPayload: DynamoFormat[NotificationPayload] = deriveDynamoFormat
   implicit val formatNotificationMessage: DynamoFormat[NotificationMessage] = deriveDynamoFormat
-  def create(
-    messageId: MessageId,
-    eventId: Long,
-    message: Message
-  ): Either[ComponentError, Option[NotificationMessage]] =
+  def create(messageId: MessageId, eventId: Long, message: Message): Either[ComponentError, List[NotificationMessage]] =
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/notifier/model/NotificationMessage.scala]:
```suggestion
  def create(resourceId: UUID, organizationId: String, eventId: Long, message: Message): Either[ComponentError, List[NotificationMessage]] =
```

File: src/main/scala/it/pagopa/interop/notifier/model/NotificationMessage.scala [hunk 4]
```diff
@@ -51,25 +52,26 @@ object NotificationMessage {
...
-            organizationId = messageId.organizationId,
-            eventId = eventId,
-            messageUUID = message.messageUUID,
-            eventJournalPersistenceId = message.eventJournalPersistenceId,
-            eventJournalSequenceNumber = message.eventJournalSequenceNumber,
-            eventTimestamp = message.eventTimestamp,
-            resourceId = messageId.resourceId.toString,
-            payload = p
-          )
+        _.fold(List[NotificationMessage]())(p =>
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/notifier/model/NotificationMessage.scala]: you can revert these changes

File: src/main/scala/it/pagopa/interop/notifier/service/impl/QueueHandler.scala [hunk 5]
```diff
@@ -50,23 +53,31 @@ final class QueueHandler(
...
     Function.unlift({ msg: Message =>
-      implicit val ctx: Seq[(String, String)] = contexts
-
+      implicit val ctx: Seq[(String, String)]                                        = contexts
       val getMessageId: PartialFunction[ProjectableEvent, Future[Option[MessageId]]] =
         PurposeEventsConverter.getMessageId(catalogManagementService, dynamoIndexService) orElse
           AgreementEventsConverter.getMessageId(dynamoIndexService) orElse
           CatalogEventsConverter.getMessageId(dynamoIndexService)
 
+      def process(nextEvent: PersistentOrganizationEvent, messageId: MessageId): Future[Unit] = {
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/notifier/service/impl/QueueHandler.scala]:
```suggestion
      def process(nextEvent: PersistentOrganizationEvent, resourceId: UUID, organizationId: String): Future[Unit] = {
```

File: src/main/scala/it/pagopa/interop/notifier/service/impl/QueueHandler.scala [hunk 6]
```diff
@@ -50,23 +53,31 @@ final class QueueHandler(
...
+        logger.debug(s"Next event id for organization ${nextEvent.organizationId} -> ${nextEvent.eventId}")
+        for {
+          notificationMessage <- NotificationMessage.create(messageId, nextEvent.eventId, msg).toFuture
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/notifier/service/impl/QueueHandler.scala]:
```suggestion
          notificationMessage <- NotificationMessage.create(resourceId, organizationId, nextEvent.eventId, msg).toFuture
```

File: src/main/scala/it/pagopa/interop/notifier/service/impl/QueueHandler.scala [hunk 7]
```diff
@@ -50,23 +53,31 @@ final class QueueHandler(
...
-              _ = logger.debug(s"Next event id for organization ${nextEvent.organizationId} -> ${nextEvent.eventId}")
-              notificationMessage <- NotificationMessage.create(messageId, nextEvent.eventId, msg).toFuture
-              ()                  <- notificationMessage.fold(Future.unit)(dynamoNotificationService.put)
-              _ = logger.debug(s"Message ${msg.messageUUID.toString} was successfully written to dynamodb")
+              events <- messageId.organizationId
+                .split(separator)
+                .toList
+                .traverse(idRetriever.getNextEventIdForOrganization)
+              _      <- events.traverse(nextEvent => process(nextEvent, messageId))
             } yield ()
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/notifier/service/impl/QueueHandler.scala]:
```
messageId.organizationId
                .split(separator)
                .toList
                .traverse {orgId =>
                idRetriever.getNextEventIdForOrganization(orgId).flatMap(nextEvent => process(nextEvent, messageId.resourceId, orgId) )
                } 
```

**Review by galales — COMMENTED** [review]
after seeing your code I realized that my suggestion was wrong, sorry 🙏 

I added few (rough) comments to ease the change, feel free to improve the code as you please

**Review by galales — APPROVED** [review]

**Review by beetlecrunch — APPROVED** [review]
